### PR TITLE
Improve S6640: Support type declarations

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Follow-up to #7290

### CODE REVIEW

File: analyzers/its/expected/Net5/Net5--net5.0-S6640.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+"location":  {
+"uri":  "sources\Net5\Net5\S4000.cs",
+"region":  {
+"startLine":  5,
+"startColumn":  12,
+"endLine":  5,
+"endColumn":  18
+}
+}
+}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/Net5/Net5--net5.0-S6640.json:28]: TPs

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/UnsafeCodeBlocks.cs [hunk 2]
```diff
@@ -46,6 +46,14 @@ protected override void Initialize(SonarAnalysisContext context)
                 }
             },
             SyntaxKind.MethodDeclaration, SyntaxKind.ConstructorDeclaration, SyntaxKind.DestructorDeclaration, SyntaxKind.OperatorDeclaration);
+        context.RegisterNodeAction(c =>
+            {
+                if (c.Node is BaseTypeDeclarationSyntax { Modifiers: var modifiers }
+                    && modifiers.Find(SyntaxKind.UnsafeKeyword) is { } unsafeModifier)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/UnsafeCodeBlocks.cs:52]: Same here. Needs a re-base.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/UnsafeCodeBlocks.cs [hunk 3]
```diff
@@ -46,6 +46,14 @@ protected override void Initialize(SonarAnalysisContext context)
...
+                {
+                    Report(c, unsafeModifier);
+                }
+            }, SyntaxKind.ClassDeclaration, SyntaxKind.InterfaceDeclaration, SyntaxKind.StructDeclaration, SyntaxKindEx.RecordClassDeclaration);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/UnsafeCodeBlocks.cs:56]: RecordStruct seems to be missing.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM.

**Pull request merged**